internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -801,37 +830,66 @@ func (r *BMACReconciler) getSpokeClient(secret *corev1.Secret) (client.Client, e
 	return r.spokeClient, err
 }
 
+// Returns a list of BMH ReconcileRequests for a given Agent
+func (r *BMACReconciler) agentToBMHReconcileRequests(ctx context.Context, agent *aiv1beta1.Agent) []reconcile.Request {
+	// No need to list all the `BareMetalHost` resources if the agent
+	// already has the reference to one of them.
+	if val, ok := agent.ObjectMeta.Labels[AGENT_BMH_LABEL]; ok {
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Namespace: agent.Namespace,
+			Name:      val,
+		}}}
+	}
+
+	bmh, err := r.findBMH(ctx, agent)
+	if bmh == nil || err != nil {
+		return []reconcile.Request{}
+	}
+
+	return []reconcile.Request{{NamespacedName: types.NamespacedName{
+		Namespace: bmh.Namespace,
+		Name:      bmh.Name,
+	}}}
+}
+
 func (r *BMACReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	mapAgentToClusterDeployment := func(a client.Object) []reconcile.Request {
+	mapAgentToBMH := func(a client.Object) []reconcile.Request {
 		ctx := context.Background()
 		agent := &aiv1beta1.Agent{}
 
 		if err := r.Get(ctx, types.NamespacedName{Name: a.GetName(), Namespace: a.GetNamespace()}, agent); err != nil {
 			return []reconcile.Request{}
 		}
 
-		// No need to list all the `BareMetalHost` resources if the agent
-		// already has the reference to one of them.
-		if val, ok := agent.ObjectMeta.Labels[AGENT_BMH_LABEL]; ok {
-			return []reconcile.Request{{NamespacedName: types.NamespacedName{
-				Namespace: agent.Namespace,
-				Name:      val,
-			}}}
+		return r.agentToBMHReconcileRequests(ctx, agent)
+	}
+
+	mapClusterDeploymentToBMH := func(a client.Object) []reconcile.Request {
+		ctx := context.Background()
+		clusterDeployment := &hivev1.ClusterDeployment{}
+
+		if err := r.Get(ctx, types.NamespacedName{Name: a.GetName(), Namespace: a.GetNamespace()}, clusterDeployment); err != nil {
+			return []reconcile.Request{}
 		}
 
-		bmh, err := r.findBMH(ctx, agent)
-		if bmh == nil || err != nil {
+		// Don't queue any reconcile if the ClusterDeployment
+		// has not been installed.
+		if !clusterDeployment.Spec.Installed {
 			return []reconcile.Request{}
 		}
 
-		return []reconcile.Request{{NamespacedName: types.NamespacedName{
-			Namespace: bmh.Namespace,
-			Name:      bmh.Name,
-		}}}
+		reconcileRequests := []reconcile.Request{}
+		agents := r.findAgentsByClusterDeployment(ctx, clusterDeployment)
+		for _, agent := range agents {
+			reconcileRequests = append(reconcileRequests, r.agentToBMHReconcileRequests(ctx, agent)...)
+		}
```

Comment:
Related to my previous comment. Another option could be to make sure that each BMH is only queued once. So, if there are multiple agents for a single BMH, we would just queue the BMH to reconcile once.

Basically, this should be a set :)

Reply:
The controller-runtime queue collapses repeat events into one, so it should work out fine anyway. But if it's easy, I agree it would be more "polite" to pass a unique list.

Reply:
I've update findAgentsByClusterDeployment to return a single Agent per BMH. This should remove any duplicate BMHs and reconcile requests.